### PR TITLE
refactor(web): convert internal prediction methods to stateless format 📚 

### PR DESCRIPTION
**Pull request by jahorton**
The previous three PRs in this sequence spun off a number of the components of the correction & prediction process into separate methods, although those methods still relied upon `ModelCompositor` state.  This PR aims to rectify that last detail, making them stateless instead.  

On the face of it, this simply means adding parameters to each method in order to re-acquire needed objects that were previously accessed through `this.` when they existed within `ModelCompositor`.  During standard use, nothing significant changes here.  What we _do_ gain, though, is quite useful for adding highly-focused unit tests:  we gain easier access to the different component methods and far simpler mocking where needed.  After all, we'll no longer need to worry about curating compositor state for tests focused on any of the spun-off methods.

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/lm-worker/src/main/model-compositor.ts [hunk 1]
```diff
@@ -166,7 +108,14 @@ export default class ModelCompositor {
...
 
-    const { postContextState, rawPredictions } = await this.correctAndEnumerate(transformDistribution, context);
+    const SEARCH_TIMEOUT = correction.SearchSpace.DEFAULT_ALLOTTED_CORRECTION_TIME_INTERVAL;
+    const timer = this.activeTimer = new correction.ExecutionTimer(this.testMode ? Number.MAX_VALUE : SEARCH_TIMEOUT, this.testMode ? Number.MAX_VALUE : SEARCH_TIMEOUT * 1.5);
+
+    const { postContextState, rawPredictions } = await correctAndEnumerate(this.contextTracker, this.lexicalModel, timer, transformDistribution, context);
+
+    if(this.activeTimer == timer) {
+      this.activeTimer = null;
+    }
```
Comment (jahorton) [inline common/web/lm-worker/src/main/model-compositor.ts:118]: Construction and state-management related to the `ExecutionTimer` used during correction-search was moved out of `correctAndEnumerate` in order to ensure `correctAndEnumerate` would be stateless.

If the goal is to be stateless, we can't have the method managing the "prediction process is active" `ModelCompositor` state signal.  This _does_ start the timer a bit earlier than usual, but the cost should easily be less than 1ms in practice; the parts between the two locations are essentially constant-time setup overhead.

File: common/web/lm-worker/src/main/model-compositor.ts [hunk 2]
```diff
@@ -181,24 +130,35 @@ export default class ModelCompositor {
...
+      deduplicatedSuggestionTuples.splice(0, ModelCompositor.MAX_SUGGESTIONS),
+      context,
+      inputTransform,
+      this.verbose
+    );
+
+    suggestions.forEach((suggestion) => {
+      suggestion.id = this.SUGGESTION_ID_SEED;
+      this.SUGGESTION_ID_SEED++;
+    });
```
Comment (jahorton) [inline common/web/lm-worker/src/main/model-compositor.ts:161]: The suggestion unique-identifier tagging process was moved out of `finalizeSuggestions` for similar reasons:  generating unique IDs is inherently stateful.

File: common/web/lm-worker/src/main/model-compositor.ts [hunk 3]
```diff
@@ -181,24 +130,35 @@ export default class ModelCompositor {
...
+    predictionAutoSelect(deduplicatedSuggestionTuples);
 
     // Section 4: Trim down the suggestion list to the N (MAX_SUGGESTIONS) most optimal,
     // then cache and return the set of suggestions
 
     // Now that we've marked the suggestion to auto-select, we can finalize the suggestions.
-    const suggestions = this.finalizeSuggestions(deduplicatedSuggestionTuples, context, inputTransform);
+    const suggestions = finalizeSuggestions(
+      this.lexicalModel,
+      deduplicatedSuggestionTuples.splice(0, ModelCompositor.MAX_SUGGESTIONS),
```
Comment (mcdurdin) [inline common/web/lm-worker/src/main/model-compositor.ts:152]: Is there a behavioural change here?

**Review by mcdurdin — APPROVED** [review]
LGTM -- have not reviewed the functions moved into the -helpers modules, assuming they are essentially unchanged

File: common/web/lm-worker/src/main/model-compositor.ts (hunk 3 above)
Comment (jahorton) [reply common/web/lm-worker/src/main/model-compositor.ts:152]: There is not.  I've just reformatted these lines and ensured that the function still gets access to properties it had been using.  As the function is no longer a member of the `ModelCompositor` class, we have to explicitly pass them now.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.75-alpha](https://keyman.com/downloads/releases/alpha/18.0.75)

File: common/web/lm-worker/src/main/model-compositor.ts (hunk 3 above)
Comment (mcdurdin) [reply common/web/lm-worker/src/main/model-compositor.ts:152]: It's the `splice` that worried me :grin:

Comment (jahorton) [reply common/web/lm-worker/src/main/model-compositor.ts:152]: Ah.  Still no - the original did a similar splice to chop off all the suggestions we wouldn't be returning to the user - should more than 12 even be found, of course.

Comment (jahorton) [reply common/web/lm-worker/src/main/model-compositor.ts:152]: From `stable-17.0`:

https://github.com/keymanapp/keyman/blob/abfc5c268fa42a52189d7eaa01571cb15b261005/common/web/lm-worker/src/main/model-compositor.ts#L435

**Pull request merged**